tensorstore/kvstore/s3/credentials/aws_credential_provider.h

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TENSORSTORE_KVSTORE_S3_AWS_CREDENTIAL_PROVIDER
+#define TENSORSTORE_KVSTORE_S3_AWS_CREDENTIAL_PROVIDER
+
+#include <functional>
+#include <string_view>
+
+#include "absl/time/time.h"
+#include "tensorstore/internal/http/http_transport.h"
+#include "tensorstore/util/result.h"
+
+namespace tensorstore {
+namespace internal_kvstore_s3 {
+
+/// Holds AWS credentials
+///
+/// Contains the access key, secret key and session token.
+/// An empty access key implies anonymous access,
+/// while the presence of a session token implies the use of
+/// short-lived STS credentials
+/// https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html
+struct AwsCredentials {
+  /// AWS_ACCESS_KEY_ID
+  std::string access_key;
+  /// AWS_SECRET_KEY_ID
+  std::string secret_key;
+  /// AWS_SESSION_TOKEN
+  std::string session_token;
+
+  bool IsAnonymous() const { return access_key.empty(); }
+};
+
+/// Base class for S3 Credential Providers
+///
+/// Implementers should override GetCredentials, IsExpired and,
+/// if the credential source supports the
+/// representation of an expiry date, ExpiresAt.
+class AwsCredentialProvider {
+  public:
+    virtual ~AwsCredentialProvider() = default;
+    virtual Result<AwsCredentials> GetCredentials() = 0;
```

Comment:
The inheritance hierarchy is mostly an attempt to align the C++ implementation with the go implementation. The `AwsCredentialProvider` interface here mostly replicates the go interface here:

https://github.com/aws/aws-sdk-go/blob/b2738d211a2ba65d26731880f30bfdb6abd80858/aws/credentials/credentials.go#L100-L108

and here:

https://github.com/aws/aws-sdk-go/blob/b2738d211a2ba65d26731880f30bfdb6abd80858/aws/credentials/credentials.go#L120-L123

As far as I understand go, the `Expirer` interface is an optional/anonymous interface that exists on the `EC2RoleProvider` [class](https://github.com/aws/aws-sdk-go/blob/b2738d211a2ba65d26731880f30bfdb6abd80858/aws/credentials/ec2rolecreds/ec2_role_provider.go#L40), but not others like the the [env](https://github.com/aws/aws-sdk-go/blob/b2738d211a2ba65d26731880f30bfdb6abd80858/aws/credentials/env_provider.go#L30-L32) provider or the [file](https://github.com/aws/aws-sdk-go/blob/b2738d211a2ba65d26731880f30bfdb6abd80858/aws/credentials/shared_credentials_provider.go#L25-L40) provider.

I can't find the `cached{chain{env, file, ec2role}}` setup in the go code now, might have another look for it in the morning. 

This is why `ExpiresAt` can return `UnimplementedError` to indicate the provider isn't handling expiry (the `CachedCredentialProvider` catches this) and `EC2CredentialProvider` inherits from `ExpiryCredentialProvider`.

Maybe `CachedCredentialProvider` shouldn't really inherit from `AwsCredentialProvider` (the go `Credentials` class/struct doesn't I think, its just a wrapper).

tl;dr. this PR aligns implementations, but maybe its overkill.

Reply:
> Maybe exposing the expiry/refresh is reasonable, but that seems more complicated than returning a Future<>, where the underlying function already checks the expiry.

How would this work? I'm a bit unsure how for example the EC2 metadata expiry date would be retrieved (and set) in the Future. I think the following illustrates my confusion:

```cpp
class EC2Prov {
  public:
    class State { ... };

    absl::Time expiry_;
    absl::Mutex mu_;
    Future<AwsCredentials> creds_; // cached

    Future<AwsCredentials> GetCredentials() {
     absl::WriterReaderLock lock(&mu_);

      if(expiry_ < absl::Now()) { // credentials have expired
        auto [promise, future] = PromiseFuturePair<AwsCredentials>(....);
        auto state = std::IntrusivePtr<State>::Make(std::move(promise));
        future.ExecuteWhenReady()[auto state = std::move(state)]() {
         ...
         auto ec2_json = ...
         expiry_ = ec2_json.ExpiryDate; // This is wrong, but illustrates my confusion
                                                          
         return AwsCredentials{ec2_json.access_key, ec2_json.secret_key, ec2_json.token};
        };
       
       creds_ = future; // Update cached credentials
     }

     return creds_; // Return credentials
   }
};

```

Reply:
In this conceptualization you might keep a shared_ptr<> to some state (could just be the expiry), then you can update that shared_ptr<> in ExecuteWhenReady returns.

Reply:
My concern is more that the expiry check `(expiry < absl::Now())`  would potentially be racy with setting `expiry_` in the `ExecuteWhenReady`. I guess it could be protected with a(nother?) mutex. Even so, in the above example, multiple calls to `GetCredentials` would end up creating multiple futures until `expiry_` was set properly.

The previous example probably isn't very good. Is there an improved way of expressing a Future with an expiry (perhaps somewhere in the codebase).

I also took a look through the `internal::oauth2` code and recalled that there's a `BearerTokenWithExpiration` idea.

https://github.com/google/tensorstore/blob/ce7eb619447818873fcad04f9b85515ad551e276/tensorstore/internal/oauth2/auth_provider.h#L26-L39

Maybe in a similar fashion an `absl::Time expiration` could be added to `AwsCredentials`? This might then 
remove the need for  `AwsCredentials::IsExpired` and `AwsCredentials::ExpiresAt`  methods and could still be used with the Chained and Caching Providers?

Reply:
I didn't figure out how implement a non-racy `Future<AwsCredentials> GetCredentials() { ... }`


> Maybe in a similar fashion an absl::Time expiration could be added to AwsCredentials? This might then
remove the need for AwsCredentials::IsExpired and AwsCredentials::ExpiresAt methods and could still be used with the Chained and Caching Providers?

I experimented with this a bit, but it ended up tending towards what was in master already. If it's desirable I could push up a PR illustrating this.